File: Bots/Python/Consumers/CodeFirst/WaterfallHostBot/dialogs/skill_http_client_listener.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import json
from typing import Dict

import aiohttp
from botbuilder.core import InvokeResponse
from botbuilder.integration.aiohttp import BotFrameworkHttpClient
from botbuilder.core.skills import (
    ConversationIdFactoryBase,
    SkillConversationIdFactoryOptions,
    BotFrameworkClient,
    BotFrameworkSkill,
)
from botbuilder.schema import (
    Activity,
    ExpectedReplies,
    ConversationReference,
    ConversationAccount,
    ChannelAccount,
    RoleTypes,
)
from botframework.connector.auth import (
    AuthenticationConstants,
    ChannelProvider,
    GovernmentConstants,
    SimpleCredentialProvider,
    MicrosoftAppCredentials,
    AppCredentials,
    MicrosoftGovernmentAppCredentials,
)

from dialogs.telemetry_listener_middleware import AppInsightsClient

class SkillHttpClientListener(BotFrameworkHttpClient):
    def __init__(
        self,
        credential_provider: SimpleCredentialProvider,
        skill_conversation_id_factory: ConversationIdFactoryBase,
        telemetry_client: AppInsightsClient,
        channel_provider: ChannelProvider = None,
    ):
        if not skill_conversation_id_factory:
            raise TypeError(
                "SkillHttpClientListener(): skill_conversation_id_factory can't be None"
            )

        super().__init__(credential_provider)

        self._skill_conversation_id_factory = skill_conversation_id_factory
        self._telemetry_client = telemetry_client
        self._channel_provider = channel_provider

    async def post_activity_to_skill(
        self,
        from_bot_id: str,
        to_skill: BotFrameworkSkill,
        service_url: str,
        activity: Activity,
        originating_audience: str = None,
    ) -> InvokeResponse:

        if originating_audience is None:
            originating_audience = (
                GovernmentConstants.TO_CHANNEL_FROM_BOT_OAUTH_SCOPE
                if self._channel_provider is not None
                and self._channel_provider.is_government()
                else AuthenticationConstants.TO_CHANNEL_FROM_BOT_OAUTH_SCOPE
            )

        options = SkillConversationIdFactoryOptions(
            from_bot_oauth_scope=originating_audience,
            from_bot_id=from_bot_id,
            activity=activity,
            bot_framework_skill=to_skill,
        )

        skill_conversation_id = await self._skill_conversation_id_factory.create_skill_conversation_id(
            options
        )
        self._telemetry_client.track_event(f'Posting activity. ActivityId: {activity.id} from BotId: {from_bot_id}')
        return await self.post_activity(
            from_bot_id,
            to_skill.app_id,
            to_skill.skill_endpoint,
            service_url,
            skill_conversation_id,
            activity,
        )

    async def post_activity(
        self,
        from_bot_id: str,
        to_bot_id: str,
        to_url: str,
        service_url: str,
        conversation_id: str,
        activity: Activity,
    ) -> InvokeResponse:
        app_credentials = await self._get_app_credentials(from_bot_id, to_bot_id)

        if not app_credentials:
            raise KeyError("Unable to get appCredentials to connect to the skill")

        # Get token for the skill call
        token = (
            app_credentials.get_access_token()
            if app_credentials.microsoft_app_id
            else None
        )

        self._telemetry_client.track_event("Init-PostActivity", 
            {'custom_dimensions': {
                'to_url': to_url, 
                'token': token,
                'activity': json.dumps(activity.as_dict(False))
                }
            }
        )

        # Capture current activity settings before changing them.
        original_conversation_id = activity.conversation.id
        original_service_url = activity.service_url
        original_relates_to = activity.relates_to
        original_recipient = activity.recipient

        try:
            activity.relates_to = ConversationReference(
                service_url=activity.service_url,
                activity_id=activity.id,
                channel_id=activity.channel_id,
                conversation=ConversationAccount(
                    id=activity.conversation.id,
                    name=activity.conversation.name,
                    conversation_type=activity.conversation.conversation_type,
                    aad_object_id=activity.conversation.aad_object_id,
                    is_group=activity.conversation.is_group,
                    role=activity.conversation.role,
                    tenant_id=activity.conversation.tenant_id,
                    properties=activity.conversation.properties,
                ),
                bot=None,
            )
            activity.conversation.id = conversation_id
            activity.service_url = service_url
            if not activity.recipient:
                activity.recipient = ChannelAccount(role=RoleTypes.skill)
            else:
                activity.recipient.role = RoleTypes.skill

            self._telemetry_client.track_event("UpdatedActivity-PostActivity", 
                {'custom_dimensions': {
                    'to_url': to_url, 
                    'token': token,
                    'activity': json.dumps(activity.as_dict(False)),
                    'json_content': json.dumps(activity.serialize())
                    }
                }
            )
            status, content = await self._post_content(to_url, token, activity)

            return InvokeResponse(status=status, body=content)

        finally:
            # Restore activity properties.
            activity.conversation.id = original_conversation_id
            activity.service_url = original_service_url
            activity.relates_to = original_relates_to
            activity.recipient = original_recipient

    async def _post_content(
        self, to_url: str, token: str, activity: Activity
    ) -> (int, object):
        headers_dict = {
            "Content-type": "application/json; charset=utf-8",
        }
        if token:
            headers_dict.update(
                {"Authorization": f"Bearer {token}",}
            )

        json_content = json.dumps(activity.serialize())

        await self._session.get(to_url.replace('/api/messages', '/api/ping?bot=WaterfallHostBotPython'))

        resp = await self._session.post(
            to_url, data=json_content.encode("utf-8"), headers=headers_dict,
        )
        resp.raise_for_status()
        data = (await resp.read()).decode()

        self._telemetry_client.track_event("PostContent-PostActivity", 
            {'custom_dimensions': {
                'to_url': to_url, 
                'token': token,
                'activity': json.dumps(activity.as_dict(False)),
                'json_content': json.dumps(activity.serialize()),
                'response_data': (await resp.read()).decode()
                }
            }
        )

        return resp.status, json.loads(data) if data else None

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Dialogs/SkillHttpClientListener.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Activity = Microsoft.Bot.Schema.Activity;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Dialogs
{
    public class SkillHttpClientListener : SkillHttpClient
    {
        private readonly SkillConversationIdFactoryBase _conversationIdFactory;
        private IBotTelemetryClient _logger;

        public SkillHttpClientListener(HttpClient httpClient, ICredentialProvider credentialProvider, SkillConversationIdFactoryBase conversationIdFactory, IChannelProvider channelProvider = null, ILogger logger = null, IBotTelemetryClient telemetryClient = null)
            : base(httpClient, credentialProvider, conversationIdFactory, channelProvider, logger)
        {
            _logger = telemetryClient;
            _conversationIdFactory = conversationIdFactory;
        }

        public virtual async Task<InvokeResponse<T>> PostActivityAsync<T>(string originatingAudience, string fromBotId, BotFrameworkSkill toSkill, Uri callbackUrl, Activity activity, CancellationToken cancellationToken)
        {
            string skillConversationId;
            try
            {
                var options = new SkillConversationIdFactoryOptions
                {
                    FromBotOAuthScope = originatingAudience,
                    FromBotId = fromBotId,
                    Activity = activity,
                    BotFrameworkSkill = toSkill
                };
                skillConversationId = await _conversationIdFactory.CreateSkillConversationIdAsync(options, cancellationToken).ConfigureAwait(false);
            }
            catch (NotImplementedException)
            {
                skillConversationId = await _conversationIdFactory.CreateSkillConversationIdAsync(activity.GetConversationReference(), cancellationToken).ConfigureAwait(false);
            }

            return await PostActivityAsync<T>(fromBotId, toSkill.AppId, toSkill.SkillEndpoint, callbackUrl, skillConversationId, activity, cancellationToken).ConfigureAwait(false);
        }

        /// <summary>
        /// Forwards an activity to a skill (bot).
        /// </summary>
        /// <param name="fromBotId">The MicrosoftAppId of the bot sending the activity.</param>
        /// <param name="toSkill">An instance of <see cref="BotFrameworkSkill"/>.</param>
        /// <param name="callbackUrl">The callback Uri.</param>
        /// <param name="activity">activity to forward.</param>
        /// <param name="cancellationToken">cancellation Token.</param>
        /// <returns>Async task with optional invokeResponse.</returns>
        public virtual async Task<InvokeResponse> PostActivityAsync(string fromBotId, BotFrameworkSkill toSkill, Uri callbackUrl, Activity activity, CancellationToken cancellationToken)
        {
            return await PostActivityAsync<object>(fromBotId, toSkill, callbackUrl, activity, cancellationToken).ConfigureAwait(false);
        }

        /// <summary>
        /// Forwards an activity to a skill (bot).
        /// </summary>
        /// <param name="fromBotId">The MicrosoftAppId of the bot sending the activity.</param>
        /// <param name="toSkill">An instance of <see cref="BotFrameworkSkill"/>.</param>
        /// <param name="callbackUrl">The callback Uri.</param>
        /// <param name="activity">activity to forward.</param>
        /// <param name="cancellationToken">cancellation Token.</param>
        /// <typeparam name="T">type of the <see cref="InvokeResponse"/> result.</typeparam>
        /// <returns>Async task with optional invokeResponse of type T.</returns>
        public virtual async Task<InvokeResponse<T>> PostActivityAsync<T>(string fromBotId, BotFrameworkSkill toSkill, Uri callbackUrl, Activity activity, CancellationToken cancellationToken)
        {
            var originatingAudience = ChannelProvider != null && ChannelProvider.IsGovernment() ? GovernmentAuthenticationConstants.ToChannelFromBotOAuthScope : AuthenticationConstants.ToChannelFromBotOAuthScope;
            return await PostActivityAsync<T>(originatingAudience, fromBotId, toSkill, callbackUrl, activity, cancellationToken).ConfigureAwait(false);
        }

        /// <summary>
        /// Forwards an activity to a skill (bot).
        /// </summary>
        /// <remarks>NOTE: Forwarding an activity to a skill will flush UserState and ConversationState changes so that skill has accurate state.</remarks>
        /// <param name="fromBotId">The MicrosoftAppId of the bot sending the activity.</param>
        /// <param name="toBotId">The MicrosoftAppId of the bot receiving the activity.</param>
        /// <param name="toUrl">The URL of the bot receiving the activity.</param>
        /// <param name="serviceUrl">The callback Url for the skill host.</param>
        /// <param name="conversationId">A conversation ID to use for the conversation with the skill.</param>
        /// <param name="activity">activity to forward.</param>
        /// <param name="cancellationToken">cancellation Token.</param>
        /// <returns>Async task with optional invokeResponse.</returns>
        public override async Task<InvokeResponse> PostActivityAsync(string fromBotId, string toBotId, Uri toUrl, Uri serviceUrl, string conversationId, Activity activity, CancellationToken cancellationToken = default)
        {
            return await PostActivityAsync<object>(fromBotId, toBotId, toUrl, serviceUrl, conversationId, activity, cancellationToken).ConfigureAwait(false);
        }

        /// <summary>
        /// Forwards an activity to a skill (bot).
        /// </summary>
        /// <remarks>NOTE: Forwarding an activity to a skill will flush UserState and ConversationState changes so that skill has accurate state.</remarks>
        /// <typeparam name="T">The type of body in the InvokeResponse.</typeparam>
        /// <param name="fromBotId">The MicrosoftAppId of the bot sending the activity.</param>
        /// <param name="toBotId">The MicrosoftAppId of the bot receiving the activity.</param>
        /// <param name="toUrl">The URL of the bot receiving the activity.</param>
        /// <param name="serviceUrl">The callback Url for the skill host.</param>
        /// <param name="conversationId">A conversation ID to use for the conversation with the skill.</param>
        /// <param name="activity">activity to forward.</param>
        /// <param name="cancellationToken">cancellation Token.</param>
        /// <returns>Async task with optional invokeResponse<typeparamref name="T"/>.</returns>
        public override async Task<InvokeResponse<T>> PostActivityAsync<T>(string fromBotId, string toBotId, Uri toUrl, Uri serviceUrl, string conversationId, Activity activity, CancellationToken cancellationToken = default)
        {
            var appCredentials = await GetAppCredentialsAsync(fromBotId, toBotId).ConfigureAwait(false);
            if (appCredentials == null)
            {
                Logger.LogError("Unable to get appCredentials to connect to the skill");
                throw new InvalidOperationException("Unable to get appCredentials to connect to the skill");
            }

            // Get token for the skill call
            var token = appCredentials == MicrosoftAppCredentials.Empty ? null : await appCredentials.GetTokenAsync().ConfigureAwait(false);

            // Clone the activity so we can modify it before sending without impacting the original object.
            var activityClone = JsonConvert.DeserializeObject<Activity>(JsonConvert.SerializeObject(activity));
            activityClone.RelatesTo = new ConversationReference
            {
                ServiceUrl = activityClone.ServiceUrl,
                ActivityId = activityClone.Id,
                ChannelId = activityClone.ChannelId,
                Locale = activityClone.Locale,
                Conversation = new ConversationAccount
                {
                    Id = activityClone.Conversation.Id,
                    Name = activityClone.Conversation.Name,
                    ConversationType = activityClone.Conversation.ConversationType,
                    AadObjectId = activityClone.Conversation.AadObjectId,
                    IsGroup = activityClone.Conversation.IsGroup,
                    Properties = activityClone.Conversation.Properties,
                    Role = activityClone.Conversation.Role,
                    TenantId = activityClone.Conversation.TenantId,
                }
            };
            activityClone.Conversation.Id = conversationId;
            activityClone.ServiceUrl = serviceUrl.ToString();
            activityClone.Recipient ??= new ChannelAccount();
            activityClone.Recipient.Role = RoleTypes.Skill;

            return await SecurePostActivityAsync<T>(toUrl, activityClone, token, cancellationToken).ConfigureAwait(false);
        }

        /// <summary>
        /// Post Activity to the bot using the bot's credentials.
        /// </summary>
        /// <param name="botId">The MicrosoftAppId of the bot.</param>
        /// <param name="botEndpoint">The URL of the bot.</param>
        /// <param name="activity">activity to post.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>InvokeResponse.</returns>
        public virtual async Task<InvokeResponse> PostActivityAsync(string botId, Uri botEndpoint, Activity activity, CancellationToken cancellationToken = default)
        {
            return await PostActivityAsync<object>(botId, botEndpoint, activity, cancellationToken).ConfigureAwait(false);
        }

        /// <summary>
        /// Post Activity to the bot using the bot's credentials.
        /// </summary>
        /// <typeparam name="T">type of invokeResponse body.</typeparam>
        /// <param name="botId">The MicrosoftAppId of the bot.</param>
        /// <param name="botEndpoint">The URL of the bot.</param>
        /// <param name="activity">activity to post.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>InvokeResponse<typeparamref name="T"/>.</returns>
        public virtual async Task<InvokeResponse<T>> PostActivityAsync<T>(string botId, Uri botEndpoint, Activity activity, CancellationToken cancellationToken = default)
        {
            // From BotId => BotId
            var appCredentials = await GetAppCredentialsAsync(botId, botId).ConfigureAwait(false);
            if (appCredentials == null)
            {
                throw new InvalidOperationException($"Unable to get appCredentials for the bot Id={botId}");
            }

            // Get token for the bot to call itself
            var token = appCredentials == MicrosoftAppCredentials.Empty ? null : await appCredentials.GetTokenAsync().ConfigureAwait(false);

            // post the activity to the url using the bot's credentials.
            Logger.LogInformation($"Posting activity. ActivityId: {activity.Id} from BotId: {botId}");
            return await SecurePostActivityAsync<T>(botEndpoint, activity, token, cancellationToken).ConfigureAwait(false);
        }

        /// <summary>
        /// Logic to build an <see cref="AppCredentials"/> object to be used to acquire tokens
        /// for this HttpClient.
        /// </summary>
        /// <param name="appId">The application id.</param>
        /// <param name="oAuthScope">The optional OAuth scope.</param>
        /// <returns>The app credentials to be used to acquire tokens.</returns>
        protected virtual async Task<AppCredentials> BuildCredentialsAsync(string appId, string oAuthScope = null)
        {
            var appPassword = await CredentialProvider.GetAppPasswordAsync(appId).ConfigureAwait(false);
            return ChannelProvider != null && ChannelProvider.IsGovernment() ? new MicrosoftGovernmentAppCredentials(appId, appPassword, HttpClient, Logger, oAuthScope) : new MicrosoftAppCredentials(appId, appPassword, HttpClient, Logger, oAuthScope);
        }

        private static T GetBodyContent<T>(string content)
        {
            try
            {
                return JsonConvert.DeserializeObject<T>(content);
            }
            catch (JsonException)
            {
                // This will only happen when the skill didn't return valid json in the content (e.g. when the status code is 500 or there's a bug in the skill)
                return default;
            }
        }

        private async Task<InvokeResponse<T>> SecurePostActivityAsync<T>(Uri toUrl, Activity activity, string token, CancellationToken cancellationToken)
        {
            _logger.TrackEvent("Init-SecurePostActivityAsync", new Dictionary<string, string>
            {
                { "toUrl", JsonConvert.SerializeObject(toUrl) },
                { "token", token },
                { "activity", JsonConvert.SerializeObject(activity) }
            });

            using (var jsonContent = new StringContent(JsonConvert.SerializeObject(activity, new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore }), Encoding.UTF8, "application/json"))
            {
                _logger.TrackEvent("Json-SecurePostActivityAsync", new Dictionary<string, string>
                {
                    { "toUrl", JsonConvert.SerializeObject(toUrl) },
                    { "token", token },
                    { "activity", JsonConvert.SerializeObject(activity) },
                    { "jsonContent", JsonConvert.SerializeObject(jsonContent) }
                });

                using (var httpRequestMessage = new HttpRequestMessage())
                {
                    httpRequestMessage.Method = HttpMethod.Post;
                    httpRequestMessage.RequestUri = toUrl;
                    if (token != null)
                    {
                        httpRequestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
                    }

                    httpRequestMessage.Content = jsonContent;

                    // // Start the child process.
                    // Process p = new Process();

                    // // Redirect the output stream of the child process.
                    // p.StartInfo.UseShellExecute = false;
                    // p.StartInfo.RedirectStandardOutput = true;
                    // p.StartInfo.FileName = "powershell.exe";
                    // p.StartInfo.Arguments = "netstat -a -n -o";
                    // p.Start();

                    // // Read the output stream first and then wait.
                    // string output = p.StandardOutput.ReadToEnd();
                    // p.WaitForExit();

                    // _logger.TrackEvent("Ports-SecurePostActivityAsync", new Dictionary<string, string>
                    //     {
                    //         { "ports", output },
                    //         { "token", token },
                    //         { "activity", JsonConvert.SerializeObject(activity) },
                    //         { "jsonContent", JsonConvert.SerializeObject(jsonContent) },
                    //         { "httpRequestMessage",  JsonConvert.SerializeObject(httpRequestMessage) },
                    //     });

                    await HttpClient.GetAsync(toUrl.ToString().Replace(toUrl.PathAndQuery, "/api/ping?bot=WaterfallHostBotDotNet"));

                    using (var response = await HttpClient.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false))
                    {
                        var responseContentAsync = string.Empty;
                        var isContentNotNull = response.Content != null;
                        if (isContentNotNull)
                        {
                            responseContentAsync = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
                        }

                        var content = isContentNotNull ? responseContentAsync : null;

                        _logger.TrackEvent("HttpClient.SendAsync-SecurePostActivityAsync", new Dictionary<string, string>
                        {
                            { "toUrl", JsonConvert.SerializeObject(toUrl) },
                            { "token", token },
                            { "activity", JsonConvert.SerializeObject(activity) },
                            { "jsonContent", JsonConvert.SerializeObject(jsonContent) },
                            { "httpRequestMessage",  JsonConvert.SerializeObject(httpRequestMessage) },
                            { "content", content },
                            { "responseContent",  JsonConvert.SerializeObject(response.Content) },
                            { "isContentNotNull",  isContentNotNull.ToString() },
                            { "responseContentAsync", responseContentAsync }
                        });

                        return new InvokeResponse<T>
                        {
                            Status = (int)response.StatusCode,
                            Body = content?.Length > 0 ? GetBodyContent<T>(content) : default
                        };
                    }
                }
            }
        }

        /// <summary>
        /// Gets the application credentials. App Credentials are cached so as to ensure we are not refreshing
        /// token every time.
        /// </summary>
        /// <param name="appId">The application identifier (AAD Id for the bot).</param>
        /// <param name="oAuthScope">The scope for the token, skills will use the Skill App Id. </param>
        /// <returns>App credentials.</returns>
        private async Task<AppCredentials> GetAppCredentialsAsync(string appId, string oAuthScope = null)
        {
            if (string.IsNullOrWhiteSpace(appId))
            {
                return MicrosoftAppCredentials.Empty;
            }

            // If the credentials are in the cache, retrieve them from there
            var cacheKey = $"{appId}{oAuthScope}";
            if (AppCredentialMapCache.TryGetValue(cacheKey, out var appCredentials))
            {
                return appCredentials;
            }

            // Credentials not found in cache, build them
            appCredentials = await BuildCredentialsAsync(appId, oAuthScope).ConfigureAwait(false);

            // Cache the credentials for later use
            AppCredentialMapCache[cacheKey] = appCredentials;
            return appCredentials;
        }
    }
}
